modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PartitionReplicaListener.java

```diff
@@ -3983,14 +3986,65 @@ private CompletableFuture<?> processOperationRequestWithTxRwCounter(
             }
         }
 
-        return processOperationRequest(senderId, request, isPrimary, opStartTsIfDirectRo, leaseStartTime)
-                .whenComplete((unused, throwable) -> {
-                    if (request instanceof ReadWriteReplicaRequest) {
-                        txRwOperationTracker.decrementOperationCount(
-                                rwTxActiveCatalogVersion(catalogService, (ReadWriteReplicaRequest) request)
-                        );
-                    }
-                });
+        UUID txIdLockingLwm = tryToLockLwmIfNeeded(request, opStartTsIfDirectRo);
+
+        try {
+            return processOperationRequest(senderId, request, isPrimary, opStartTsIfDirectRo, leaseStartTime)
+                    .whenComplete((unused, throwable) -> {
+                        unlockLwmIfNeeded(txIdLockingLwm, request);
+
+                        if (request instanceof ReadWriteReplicaRequest) {
+                            txRwOperationTracker.decrementOperationCount(
+                                    rwTxActiveCatalogVersion(catalogService, (ReadWriteReplicaRequest) request)
+                            );
+                        }
+                    });
+        } catch (Throwable e) {
+            try {
+                unlockLwmIfNeeded(txIdLockingLwm, request);
+            } catch (Throwable unlockProblem) {
+                e.addSuppressed(unlockProblem);
+            }
+            throw e;
+        }
+    }
+
+    private static UUID newFakeTxId() {
+        return UUID.randomUUID();
+    }
+
+    private @Nullable UUID tryToLockLwmIfNeeded(ReplicaRequest request, @Nullable HybridTimestamp opStartTsIfDirectRo) {
```

Comment:
Could you please add a javadoc that will explain when we do lock the LWM and when we unlock it? Basically, explanation that you have in PR, seems perfect here.

> When executing an operation in an RO transaction (explicit or implicit), attempt to lock LWM on the data node where it's executed
If lock attempt fails, throw an exception with a specific error code
When cleaning up after an RO transaction had been finished, unlock LWM on each node where such a cleanup happens
For direct RO operations (which happen in implicit RO transactions), unlock LWM right after the read has been done on the data node